perf_bench_test.go

```diff
@@ -24,127 +24,619 @@ package main
+	require.NoError(t, err, "Failed writting file")
+	writer.Flush()
 
-	log.Debugf("vmID %s: returned in %f seconds", vmIDString, time.Since(start).Seconds())
+	for _, metric := range metrics {
+		var data []string
+		for _, header := range headers {
+			value, isPresent := metric[header]
+			if isPresent {
+				vStr := strconv.FormatFloat(value, 'f', -1, 64)
+				data = append(data, vStr)
+			} else {
+				data = append(data, "")
+			}
+		}
+		err = writer.Write(data)
+		require.NoError(t, err, "Failed writting file")
+		writer.Flush()
+	}
 }
 
-func getImages() map[string]string {
-	return map[string]string{
-		"helloworld": "vhiveease/helloworld:var_workload",
-		// "chameleon":    "vhiveease/chameleon:var_workload",
-		// "pyaes":        "vhiveease/pyaes:var_workload",
-		// "image_rotate": "vhiveease/image_rotate:var_workload",
-		// "json_serdes":  "vhiveease/json_serdes:var_workload",
-		// "lr_serving":   "vhiveease/lr_serving:var_workload",
-		// "cnn_serving":  "vhiveease/cnn_serving:var_workload",
-		// "rnn_serving":  "vhiveease/rnn_serving:var_workload",
-		// "lr_training":  "vhiveease/lr_training:var_workload",
+// getCPUIntenseRPS returns the number of requests per second that stress CPU for each image.
+func getCloseLoopRPS() int {
+	var (
+		sum, result int
+		values      []int
+		funcs       = strings.Split(*funcNames, ",")
+		reqsPerSec  = map[string]int{
+			"helloworld":   1000,
+			"chameleon":    85,
+			"pyaes":        1000,
+			"image_rotate": 333,
+			"json_serdes":  167,
+			"lr_serving":   1000,
+			"cnn_serving":  20,
+			"rnn_serving":  100,
+		}
+	)
+
+	for _, funcName := range funcs {
+		values = append(values, reqsPerSec[funcName])
+		sum += reqsPerSec[funcName]
+	}
+
+	for _, rps := range values {
+		result += rps * rps / sum
 	}
+
+	return result
+}
+
+func getUnloadedServiceTime() float64 {
+	var (
+		sum         float64
+		funcs       = strings.Split(*funcNames, ",")
+		serviceTime = map[string]float64{
+			"helloworld":   1,
+			"chameleon":    12,
+			"pyaes":        1,
+			"image_rotate": 3,
+			"json_serdes":  6,
+			"lr_serving":   1,
+			"cnn_serving":  60,
+			"rnn_serving":  10,
+		}
+	)
+
+	for _, funcName := range funcs {
+		sum += serviceTime[funcName]
+	}
+
+	return sum / float64(len(funcs))
+}
+
+func calculateRPS(vmNum int) int {
+	baseRPS := getCloseLoopRPS()
+	return vmNum * baseRPS
+}
+
+func bindSocket() error {
+	var (
+		coreFree = true
+		cores    = 32
+	)
+	pidBytes, err := getFirecrackerPid()
+	if err != nil {
+		return err
+	}
+
+	vmPidList := strings.Split(string(pidBytes), " ")
+	for _, vm := range vmPidList {
+		vm = strings.TrimSpace(vm)
+		if cores > len(vmPidList) {
+			if coreFree {
+				log.Debugf("binding pid: %s to core 13", vm)
+				if err := bindProcessToCPU("13", vm); err != nil {
+					return err
+				}
+				coreFree = false
+			} else {
+				log.Debugf("binding pid: %s", vm)
+				if err := bindProcessToCPU("1-11:2,15-43:2,47-63:2", vm); err != nil {
```

Comment:
core-to-socket affinity may be different on different CPUs.

Reply:
Done